cpp/src/strings/convert/convert_urls.cu

```diff
@@ -185,137 +192,184 @@ constexpr bool is_hex_digit(char ch)
+  __shared__ typename cub::WarpReduce<int8_t>::TempStorage cub_storage[warps_per_threadblock];
+
+  int global_thread_id  = blockIdx.x * blockDim.x + threadIdx.x;
+  int global_warp_id    = global_thread_id / cudf::detail::warp_size;
+  int local_warp_id     = threadIdx.x / cudf::detail::warp_size;
+  int warp_lane         = threadIdx.x % cudf::detail::warp_size;
+  int nwarps            = gridDim.x * blockDim.x / cudf::detail::warp_size;
+  char* in_chars_shared = temporary_buffer[local_warp_id];
+
+  // Loop through strings, and assign each string to a warp.
+  for (size_type row_idx = global_warp_id; row_idx < num_strings; row_idx += nwarps) {
+    auto in_chars_string          = in_chars + in_offsets[row_idx];
+    auto string_length            = in_offsets[row_idx + 1] - in_offsets[row_idx];
+    int nblocks                   = (string_length + block_size - 1) / block_size;
+    offset_type escape_char_count = 0;
+
+    for (int block_idx = 0; block_idx < nblocks; block_idx++) {
+      int string_length_block = std::min(block_size, string_length - block_size * block_idx);
+
+      // Each warp collectively loads input characters of the current block to the shared memory.
+      // Two halo cells with 0s are added after the input characters to avoid branches when testing
+      // escape characters.
+      for (int char_idx = warp_lane; char_idx < string_length_block + 2;
+           char_idx += cudf::detail::warp_size) {
+        char ch    = 0;
+        int in_idx = block_idx * block_size + char_idx;
+        if (in_idx < string_length) ch = in_chars_string[in_idx];
+        in_chars_shared[char_idx] = ch;
+      }
+
+      __syncwarp();
+
+      // `char_idx_start` represents the start character index of the current warp.
+      for (int char_idx_start = 0; char_idx_start < string_length_block;
+           char_idx_start += cudf::detail::warp_size) {
+        int char_idx = char_idx_start + warp_lane;
+        int8_t is_ichar_escape_char =
+          (char_idx < string_length_block && is_escape_char(in_chars_shared + char_idx)) ? 1 : 0;
+
+        // Warp-wise reduction to calculate the number of escape characters.
+        // All threads in the warp participate in the reduction, even if `char_idx` is beyond
+        // `string_length_block`.
+        int8_t total_escape_char =
+          cub::WarpReduce<int8_t>(cub_storage[local_warp_id]).Sum(is_ichar_escape_char);
+
+        if (warp_lane == 0) { escape_char_count += total_escape_char; }
+
+        __syncwarp();
+      }
+    }
+    // URL decoding replaces 3 bytes with 1 for each escape character.
+    if (warp_lane == 0) { out_counts[row_idx] = string_length - escape_char_count * 2; }
   }
-};
+}
 
 /**
- * @brief Functor for replacing character escape sequences in URL-encoded strings.
+ * @brief Decode and copy from the input char column to the output char column.
+ *
+ * @tparam threadblock_size Number of threads in a threadblock. This template argument must be a
```

Comment:
Most of the comments from the previous kernel also apply to this one:
- Renaming the template parameters
- Using more unsigned ints
- Maybe an extra warp sync
- Using a constant for the halo size (4 instead of 2 in this kernel).

Reply:
Addressed some of these points in https://github.com/rapidsai/cudf/pull/8622/commits/961ed39a28b0bcac1f2d4abcc2b6855309535843, https://github.com/rapidsai/cudf/pull/8622/commits/1505e15e51f9f8a4d5bc7d6119a35a5ebe163c88, https://github.com/rapidsai/cudf/pull/8622/commits/161468fdf6c4415d7990f887e8429c9e34800ec0. Let's use the other thread to discuss unsigned int.